crates/client/rpc/src/lib.rs

```diff
@@ -1433,18 +1425,107 @@ where
         &self,
         transaction_hash: FieldElement,
     ) -> RpcResult<MaybePendingTransactionReceipt> {
-        let substrate_block_hash = self
-            .backend
-            .mapping()
-            .block_hash_from_transaction_hash(transaction_hash.into())
-            .map_err(|e| {
+        let chain_id = self.chain_id()?.0.into();
+        let receipt =
+            match self.backend.mapping().block_hash_from_transaction_hash(transaction_hash.into()).map_err(|e| {
                 error!("Failed to interact with db backend error: {e}");
                 StarknetRpcApiError::InternalServerError
-            })?
-            .ok_or(StarknetRpcApiError::TxnHashNotFound)?;
+            })? {
+                Some(substrate_block_hash) => {
+                    self.prepare_tx_receipt(chain_id, transaction_hash, substrate_block_hash).await?
+                }
+                // Try to find pending Tx
+                None => self.get_pending_transaction_receipt(chain_id, transaction_hash).await.map_err(|e| {
+                    error!("Failed to find pending tx with hash: {transaction_hash}: {e}");
+                    StarknetRpcApiError::TxnHashNotFound
+                })?,
+            };
+        Ok(receipt)
+    }
+}
 
-        let starknet_block: mp_block::Block = get_block_by_block_hash(self.client.as_ref(), substrate_block_hash)?;
+/// RPC Helper methods
+impl<A, B, BE, G, C, P, H> Starknet<A, B, BE, G, C, P, H>
+where
+    A: ChainApi<Block = B> + 'static,
+    B: BlockT,
+    P: TransactionPool<Block = B> + 'static,
+    BE: Backend<B> + 'static,
+    C: HeaderBackend<B> + BlockBackend<B> + StorageProvider<B, BE> + 'static,
+    C: ProvideRuntimeApi<B>,
+    C::Api: StarknetRuntimeApi<B> + ConvertTransactionRuntimeApi<B>,
+    G: GenesisProvider + Send + Sync + 'static,
+    H: HasherT + Send + Sync + 'static,
+{
+    fn prepare_pending_block_with_tx_hashes(
+        &self,
+        chain_id: Felt252Wrapper,
+    ) -> Result<PendingBlockWithTxHashes, StarknetRpcApiError> {
+        let parent_hash = self.client.info().best_hash;
+        let latest_block = get_block_by_block_hash(self.client.as_ref(), parent_hash).unwrap_or_default();
+        let latest_block_header = latest_block.header();
+        let transactions = self
+            .get_pending_txs(parent_hash)?
+            .iter()
+            .map(|tx| tx.compute_hash::<H>(chain_id.0.into(), false).0)
```

Comment:
let's use the cache here if possible

Reply:
Yeah, maybe it's prematurate to store the hash of tx that may never be actually executed.
Even if they are ready, they may end up being invalid and not making it into the block (for old tx version, the new one will just be reverted but there)